### PR TITLE
feat(android/app): Provide haptic feedback when typing

### LINKED ISSUE #5792: feat(android): vibration while typing
I support a user that would like the option of having the device vibrate when typing on the screen. This haptic feedback would be helpful.

### PR DESCRIPTION
**Pull request by darcywong00**
Fixes #5792

This feature adds a preference to the Keyman Settings menu for the user to control if the Keyman keyboard provides haptic feedback (vibrates). The default for this preference is off.
settings-screen-ap

This initial implementation adds haptic feedback when the key inserts text.

## User Testing
Setup - 
1. Install the PR build of "Keyman for Android" on a physical Android device. Cannot use Android emulator because you'll need to be able to feel the device's haptic feedback (vibrate)

* **TEST_HAPTIC_FEEDBACK** - Tests haptic feedback controlled by preference toggle
1. Start Keyman for Android
2. On the "Get Started" menu, enable Keyman as a system keyboard and set as default system keyboard.
3. Dismiss the "Get Started" menu.
4. With the default sil_euro_latin keyboard, type in the Keyman app
5. Verify as text is output, the keyboard does not vibrate
6. In the Keyman Settings menu, find the preference for "Vibrate when typing"
7. Verify the preference default is off
8. Enable the "Vibrate when typing" preference and exit the menu
9. With the default sil_euro_latin keyboard, type in the Keyman app
10. Verify as text is output, the keyboard provides haptic feedback (vibrates)
11. Close the Keyman app and launch a separate app (e.g. Google Chrome)
12. Select a field to enter text and have the Keyman keyboard displayed as a system keyboard
13. With the default sil_euro_latin keyboard, type
14. Verify as text is output, the keyboard provides haptic feedback (vibrates)

### CODE REVIEW

**Comment by bharanidharanj** [comment]
* **TEST_HAPTIC_FEEDBACK (PASSED):** Tested this (as per the instructions) with the attached PR build (Keyman 15.0.248-beta-test-6626) in my Android Mobile device (Version 11) and it is working as expected. ie., the mobile vibrates on each keypress in Keyman In-App as well as using System Keyboard. (eg., Chrome browser)

File: android/KMAPro/kMAPro/src/main/java/com/keyman/android/SystemKeyboard.java [hunk 1]
```diff
@@ -69,6 +69,9 @@ public void onCreate() {
     KMManager.SpacebarText spacebarText = KMManager.SpacebarText.fromString(prefs.getString(KeymanSettingsActivity.spacebarTextKey, KMManager.SpacebarText.LANGUAGE_KEYBOARD.toString()));
     KMManager.setSpacebarText(spacebarText);
 
+    boolean mayVibrateWhenTyping = prefs.getBoolean(KeymanSettingsActivity.vibrateWhenTypingKey, false);
+    KMManager.setMayVibrateWhenTyping(mayVibrateWhenTyping);
```
Comment (mcdurdin) [inline android/KMAPro/kMAPro/src/main/java/com/keyman/android/SystemKeyboard.java:73]: It'd be good if we used a consistent name for the variable and the setting throughout, e.g. `hapticFeedback` and `hapticFeedbackKey`.

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsFragment.java [hunk 2]
```diff
@@ -149,6 +149,21 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
       }
     });
 
+    SwitchPreference hapticPreference = new SwitchPreference(context);
```
Comment (mcdurdin) [inline android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsFragment.java:152]: Suggest `hapticFeedbackPreference`, etc.

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java [hunk 3]
```diff
@@ -743,6 +745,12 @@ private void checkSendCrashReport() {
     KMManager.setMaySendCrashReport(maySendCrashReport);
   }
 
+  private void checkMayVibrateWhenTyping() {
```
Comment (mcdurdin) [inline android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java:748]: Ditto on consistent naming, e.g. `checkHapticFeedback()`

File: android/KMAPro/kMAPro/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -135,6 +135,9 @@
   <!-- Context: Keyman Settings menu / Spacebar caption hint -->
   <string name="spacebar_caption_hint_blank" comment="Spacebar caption hint - blank">Do not show caption on spacebar</string>
 
+  <!-- Context: Keyman Settings menu / Haptic feedback -->
+  <string name="haptic_feedback" comment="haptic feedback on key press">Vibrate when typing</string>
```
Comment (mcdurdin) [inline android/KMAPro/kMAPro/src/main/res/values/strings.xml:139]: It's fine for the internal name to be `haptic_feedback` and the user-visible text to be "Vibrate when typing" :grin:

**Comment by darcywong00** [comment]
>  but not sure it really fits into beta because it's a new feature. Thoughts on that?

Yeah, I should get my wrist slapped for doing a feature this late in beta. But as I looked for low-hanging fruit on the Android issues, it seemed this was small enough and a good UX improvement for @rmlockwood

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsActivity.java [hunk 5]
```diff
@@ -13,6 +13,7 @@ public class KeymanSettingsActivity extends BaseActivity {
   protected static final String showBannerKey = "ShowBanner";
   protected static final String sendCrashReport = "SendCrashReport";
   public static final String spacebarTextKey = "SpacebarText";
+  public static final String hapticFeedbackKey = "HapticFeedbackKey";
```
Comment (mcdurdin) [inline android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/KeymanSettingsActivity.java:16]:
```suggestion
  public static final String hapticFeedbackKey = "HapticFeedback";
```

File: android/help/basic/config/index.md [hunk 6]
```diff
@@ -47,6 +47,9 @@ Click on this to [adjust the keyboard height](adjust-keyboard-height).
 ## Spacebar Caption
 Click on this to [change the displayed keyboard name](spacebar-caption) on the spacebar.
 
+## Vibrate When Typing
+When enabled, the Keyman keyboard will provide haptic feedback (vibrate) when the user presses a key and generates output. The default preference is off.
```
Comment (mcdurdin) [inline android/help/basic/config/index.md:51]: Question: is there an Android system preference for this as well? Should Keyman follow the Android system preference? Can you outline the various possible modes?

Comment (darcywong00) [reply android/help/basic/config/index.md:51]: On my Android 11 device, I don't see a system preference.
But Gboard has this under "Preferences"
* Toggle for "Haptic feedback on keypress"
* Slider for "Vibration strength on keypress"
    * "System default" is 0ms
    * Max is 100 ms

**Review by mcdurdin — APPROVED** [review]
OK, LGTM. We'll squeeze it in to beta I guess!

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.250-beta](https://keyman.com/downloads/releases/beta/15.0.250)

**Comment by khashashin** [comment]
@darcywong00 & @mcdurdin thank you for that feature! How could I implement it on my System Keyboard that I've developed using KeymanEngine?

I just created checkbox inside the application and not sure how to implement the logic behind that.

image

image

```java
package ce.latin.keyboard;

import androidx.appcompat.app.AppCompatActivity;
import android.content.Context;
import android.content.Intent;
import android.provider.Settings;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.CheckBox;

public class MainActivity extends AppCompatActivity {

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    setTheme(R.style.AppTheme);
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    Button button1 = (Button) findViewById(R.id.button1);
    button1.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        startActivity(new Intent(Settings.ACTION_INPUT_METHOD_SETTINGS));
      }
    });

    Button button2 = (Button) findViewById(R.id.button2);
    button2.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        InputMethodManager imManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        imManager.showInputMethodPicker();
      }
    });

    CheckBox toggleVibrate = (CheckBox) findViewById(R.id.toggleVibrateOnTyping);
    toggleVibrate.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {

      }
    });
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    // Inflate the menu; this adds items to the action bar if it is present.
    getMenuInflater().inflate(R.menu.menu_main, menu);
    return true;
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    // Handle action bar item clicks here. The action bar will
    // automatically handle clicks on the Home/Up button, so long
    // as you specify a parent activity in AndroidManifest.xml.
    int id = item.getItemId();

    //noinspection SimplifiableIfStatement
    if (id == R.id.action_settings) {
      return true;
    }

    return super.onOptionsItemSelected(item);
  }
}

```

**Pull request merged**